src/Shared/FileDelegates.cs

```diff
@@ -6,13 +6,22 @@
 namespace Microsoft.Build.Shared
 {
     /// <summary>
-    /// delegate for System.IO.Directory.GetFiles, used for testing
+    /// delegate for System.IO.Directory.GetFiles
     /// </summary>
     /// <param name="path">Directory path to start search for files in</param>
     /// <param name="searchPattern">pattern of files to match</param>
     /// <returns>string array of files which match search pattern</returns>
     internal delegate string[] DirectoryGetFiles(string path, string searchPattern);
 
+    /// <summary>
+    /// delegate for optimized looking up files in directory
+    /// designed to be used for testing and to verify file existence by directory listing
+    /// </summary>
+    /// <param name="path">Directory path to start search for files in</param>
+    /// <param name="fileName">name of file</param>
+    /// <returns>return full path file name or null if file do not exists</returns>
+    internal delegate string DirectoryFile(string path, string fileName);
```

Comment:
minor: why a custom delegate instead of ```Func<string, string, string>```?

Reply:
That's how it was previously...though honestly, I do like the idea of a Func<string, string, string>

Reply:
I was trying to be consistent with existing code. 
I am not used to use delegates for DI purposes and have not wrapped my head around it yet, given my past experience I'd rather use interfaces here, but as for readability in this particular case using predefined delegate is more readable - to me.